quic/s2n-quic-core/src/sync/worker.rs

```diff
@@ -0,0 +1,217 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::sync::primitive::{Arc, AtomicUsize, AtomicWaker, Ordering};
+use cache_padded::CachePadded;
+use core::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// Creates a worker channel with a Sender and Receiver
+pub fn channel() -> (Sender, Receiver) {
+    let state = Arc::new(State::default());
+    let sender = Sender(state.clone());
+    let receiver = Receiver { state, credits: 0 };
+    (sender, receiver)
+}
+
+/// A handle to the receiver side of the worker channel
+///
+/// This handle is used by the worker to wake up when there is work to do.
+pub struct Receiver {
+    state: Arc<State>,
+    credits: usize,
+}
+
+impl Receiver {
+    /// Acquires work to be processed for the Receiver
+    ///
+    /// `None` is returned when there are no more active Senders.
+    #[inline]
+    pub async fn acquire(&mut self) -> Option<usize> {
+        Acquire(self).await
+    }
+
+    /// Polls work to be processed for the receiver
+    ///
+    /// `None` is returned when there are no more active Senders.
+    #[inline]
+    pub fn poll_acquire(&mut self, cx: &mut Context) -> Poll<Option<usize>> {
+        let state = &*self.state;
+
+        macro_rules! acquire {
+            () => {{
+                // take the credits that we've been given by the senders
+                self.credits += state.remaining.swap(0, Ordering::Acquire);
+
+                // if we have any credits then return
+                if self.credits > 0 {
+                    return Poll::Ready(Some(self.credits));
+                }
+            }};
+        }
+
+        // first try to acquire credits
+        acquire!();
+
+        // if we didn't get any credits then register the waker
+        state.receiver.register(cx.waker());
+
+        // make one last effort to acquire credits in case a sender submitted some while we were
+        // registering the waker
+        acquire!();
```

Comment:
What happens if we dont call this? We would just wake up right? Should we de-register the waker if we were able to acquire credits?

Reply:
spoke offline: we need to check otherwise it possible to deadlock. sender tried waking receive but since it wasnt registered, it will never wake.

It not critical to de-register since we will simply wake up pre-maturely at worse, but thats not expected to happen or be detrimental.